homeassistant/components/tesla_wall_connector/__init__.py

```diff
@@ -0,0 +1,185 @@
+"""The Tesla Wall Connector integration."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import timedelta
+import logging
+from typing import Any, Callable
+
+from tesla_wall_connector import WallConnector
+from tesla_wall_connector.exceptions import (
+    WallConnectorConnectionError,
+    WallConnectorConnectionTimeoutError,
+    WallConnectorError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    CONF_SCAN_INTERVAL_CHARGING,
+    DEFAULT_SCAN_INTERVAL,
+    DEFAULT_SCAN_INTERVAL_CHARGING,
+    DOMAIN,
+    WALLCONNECTOR_DATA_LIFETIME,
+    WALLCONNECTOR_DATA_VITALS,
+    WALLCONNECTOR_DEVICE_NAME,
+)
+
+PLATFORMS: list[str] = ["binary_sensor"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Tesla Wall Connector from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    hostname = entry.data[CONF_HOST]
+    poll_interval = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    poll_interval_charging = entry.options.get(
+        CONF_SCAN_INTERVAL_CHARGING, DEFAULT_SCAN_INTERVAL_CHARGING
+    )
+
+    wall_connector = WallConnector(host=hostname, session=async_get_clientsession(hass))
+
+    try:
+        version_data = await wall_connector.async_get_version()
+    except WallConnectorError as ex:
+        raise ConfigEntryNotReady from ex
+
+    coordinator: DataUpdateCoordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="tesla-wallconnector",
+        update_interval=timedelta(seconds=poll_interval),
+    )
+
```

Comment:
Setup the coordinator and call `await coordinator.async_config_entry_first_refresh()`. It will take care of raising `ConfigEntryNotReady` if the update fails.

Then you can drop 
```
    try:
        version_data = await wall_connector.async_get_version()
    except WallConnectorError as ex:
        raise ConfigEntryNotReady from ex
```

Reply:
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/update_coordinator.py#L149

Reply:
We use the get_version() call to retrieve and store the serial number and other info in the WallConnectorData object, therefore it cannot be dropped..

Reply:
Can you instead move the `await coordinator.async_config_entry_first_refresh()` higher before data is stored in `hass.data` so in case it raises we haven't stored anything yet?

Reply:
Moved storing of the data until after async_config_entry_first_refresh() 24de889